neo/Ledger/MemPool.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Neo.Network.P2P.Payloads;
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+
+namespace Neo.Ledger
+{
+    internal class MemPool : IReadOnlyCollection<Transaction>
+    {
+        private class PoolItem
+        {
+            public readonly Transaction Transaction;
+            public readonly DateTime Timestamp;
+
+            public PoolItem(Transaction tx)
+            {
+                Transaction = tx;
+                Timestamp = DateTime.UtcNow;

```

Comment:
As @erikzhang commented, transactions gonna expire and, perhaps, RPC seeds nodes will relay it again.
Malicious RPC nodes can even run a loop for relaying the same transaction because the DateTime.UtcNow is given by the time that each node receives, right @jsolman and @shargon ?

Another option (but far from our reality) would be to sign the time of the transaction...aehauheauheauea But I think it would be too much for now.

Reply:
Yes, I mentioned early along in this PR that this won’t prevent the transactions from getting back in the mem_pool. It can potentially help some old ones die out, but no guarantee even without someone purposely relaying them.

Free transactions will probably eventually need to have a proof of work to order them; even trying to sort them by some random lottery will still really amount to a proof of work since the algorithm will be known.

Free transactions that need to guarantee they aren’t in the mempool for too long should protect themselves from being stuck in the pool for too long by including something to make them invalid after a certain number of blocks. This can be done today already for invocation transactions. Enforcing Signing the time of all transactions by the sender really isn’t the right answer anyway for how to order free transactions, because an old transaction arguably should have higher priority than a newer one; it has been waiting longer after all and they are both free. Keep in mind if the transaction gets old it is because of a high amount of free transactions causing some to not be accepted due to the 20 free per block minimum. Having it get old doesn’t mean it is any less valid than a new one that just got relayed. So, as stated if the transaction should be become invalid after a period of not being relayed it is the responsibility of the creator of the transaction to create it such that it fails to verify after ‘x’ blocks; otherwise it should only get ejected if it could not fit in the mempool any longer — and for that it is dependent on how they are sorted, which today for free transactions is actually based on the one with the smaller hash; and since one could ad an arbitrary random attribute and keep creating transactions to get a smaller hash; it is already ordered by proof of work today for free transactions; it’s just not widely known.

So in conclusion; while I like making the MemPool it’s own class; I’m not really that enthusiastic about removing free transactions just for being old.

Sometimes someone might want to cancel their transaction that was free if it is getting old and it is tying up one of their utxo’s. For that I think it would be nice if they could sign a transaction requesting one of their previously free transactions be removed from mempools, and that be relayed around the network while nodes remove the transaction. That would allow them to send a new transaction with a fee and untie their utxo. That is the only real value I see to this business of removing an old free transaction (for the case it is tying up someone’s old utxo from being used again).

Reply:
Hi, @jsolman, please, let's not even consider proof-of-work here. There is no reason for that.
Several other strategies can be applied, I prefer even use Learning techniques than using and PoW concept.

Previously, I had lost this part: "otherwise it should only get ejected if it could not fit in the mempool any longer"
It is important to take that is mind.

Great insights. I agree.
But I believe that something can still be designed, even with the idea of "address" signing the timestamp when making the request. But let the time tell the needs. heauheauae